# Домашнє завдання до теми "Розширені методи обробки даних у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd
from typing import Union


data_path = "data/supermarket_sales.csv"
supermarket_dataset = pd.read_csv(data_path)

supermarket_dataset.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4  
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3

1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [6]:
supermarket_dataset.sort_values(by="Unit price", ascending=False).head()

Invoice ID Branch       City Customer type Gender         Product line  \
983  148-41-7930      C  Naypyitaw        Normal   Male    Health and beauty   
122  219-22-9386      B   Mandalay        Member   Male    Sports and travel   
930  641-62-7288      B   Mandalay        Normal   Male   Home and lifestyle   
494  437-53-3084      B   Mandalay        Normal   Male  Fashion accessories   
283  667-92-0055      A     Yangon        Member   Male    Health and beauty   

     Unit price  Quantity  Tax 5%    Total       Date   Time      Payment  \
983       99.96         7  34.986  734.706  1/23/2019  10:33         Cash   
122       99.96         9  44.982  944.622   3/9/2019  17:26  Credit card   
930       99.92         6  29.976  629.496  3/24/2019  13:33      Ewallet   
494       99.89         2   9.989  209.769  2/26/2019  11:48      Ewallet   
283       99.83         6  29.949  628.929   3/4/2019  15:02      Ewallet   

       cogs  gross margin percentage  gross income  Rating Rating Category  
983  699.72                 4.761905        34.986     6.1          Medium  
122  899.64                 4.761905        44.982     4.2             Low  
930  599.52                 4.761905        29.976     7.1          Medium  
494  199.78                 4.761905         9.989     7.1          Medium  
283  598.98                 4.761905        29.949     8.5            High

2. Знайдіть інвойс (рядок в даних) в категорії товарів Electronic accessories, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

In [7]:
supermarket_dataset[supermarket_dataset["Product line"]=="Electronic accessories"].agg(["min", "max"])

Invoice ID Branch      City Customer type  Gender  \
min  114-35-5271      A  Mandalay        Member  Female   
max  891-01-7034      C    Yangon        Normal    Male   

               Product line  Unit price  Quantity   Tax 5%     Total  \
min  Electronic accessories       10.56         1   1.2725   26.7225   
max  Electronic accessories       99.73        10  44.8785  942.4485   

         Date   Time  Payment    cogs  gross margin percentage  gross income  \
min  1/1/2019  10:02     Cash   25.45                 4.761905        1.2725   
max  3/9/2019  20:59  Ewallet  897.57                 4.761905       44.8785   

     Rating Rating Category  
min     4.0            High  
max    10.0          Medium

3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу: високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 до 8 не включаючи - то він середній (`Medium`)
- інакше - низький (`Low`).

In [8]:
def category_mapping(value: Union[int, float]) -> str:
    """
    Map a numerical value to a categorical label based on predefined ranges.

    Parameters
    ----------
    value : Union[int, float]
        The numerical value to categorize.

    Returns
    -------
    str
        The categorical label for the given value. Possible values are:
        - 'High' for values >= 8
        - 'Medium' for values >= 6 and < 8
        - 'Low' for values < 6

    Examples
    --------
    >>> category_mapping(9)
    'High'

    >>> category_mapping(7)
    'Medium'

    >>> category_mapping(5)
    'Low'
    """
    if value >= 8:
        return 'High'
    elif value >= 6:
        return 'Medium'
    else:
        return 'Low'


In [9]:
supermarket_dataset['Rating Category'] = supermarket_dataset.Rating.apply(category_mapping)
supermarket_dataset.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  \
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1   
1  10:29         Cash   76.40                 4.761905        3.8200     9.6   
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4   
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4   
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3   

  Rating Category  
0            High  
1            High  
2          Medium  
3            High  
4             Low

In [10]:
supermarket_dataset['Rating Category'] = supermarket_dataset['Rating Category'].astype('category')

3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [11]:
rating_counts = supermarket_dataset["Rating Category"].value_counts()
rating_counts/rating_counts.sum() * 100

Rating Category
Medium    35.6
High      32.9
Low       31.5
Name: count, dtype: float64

4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [12]:
time = '11:28'

In [13]:
def get_hour(time: str) -> int:
    """
    Extract the hour from a time string in the format "HH:MM:SS" or "HH:MM".

    Parameters
    ----------
    time : str
        The time string in the format "HH:MM:SS" or "HH:MM".

    Returns
    -------
    int
        The hour extracted from the time string as an integer.

    Examples
    --------
    >>> get_hour("14:30:00")
    14

    >>> get_hour("08:45")
    8
    """
    return int(time.split(':')[0])

get_hour(time)

11

4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

In [14]:
supermarket_dataset['Hour'] = supermarket_dataset.Time.apply(get_hour)
supermarket_dataset.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  \
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1   
1  10:29         Cash   76.40                 4.761905        3.8200     9.6   
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4   
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4   
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3   

  Rating Category  Hour  
0            High    13  
1            High    10  
2          Medium    13  
3            High    20  
4             Low    10

4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому на Ваш погляд саме о цих годинах? Чи це слогічно, що найбільше покупок саме в ці години?".

In [17]:
supermarket_dataset['Hour'].value_counts()

Hour
19    113
13    103
15    102
10    101
18     93
11     90
12     89
14     83
16     77
20     75
17     74
Name: count, dtype: int64

<p style='font-family: Garamond ; font-size: 18px;'>Розглянемо <b>топ-5</b> продуктивних годин пік щодо активу ЦА мережі магазинів нашого клієнта.</p>

<p style='font-family: Garamond ; font-size: 18px;'>Маємо наступний ранжований ряд:</p>
<ol style='font-family: Garamond ; font-size: 18px;'>
    <li>19:00 (7 PM).</li>
    <li>13:00 (1 PM).</li>
    <li>15:00 (3 PM).</li>
    <li>10:00 (10 AM).</li>
    <li>18:00 (6 PM).</li>
</ol>
<hr></hr>
<ul style='font-family: Garamond ; font-size: 18px;'>
    <li><b>19:00</b></li>
    <ul style='font-family: Garamond ; font-size: 18px;'>Цілком логічна тенденція, оскільки 19:00 - класичне завершення робочого дня за фіксованим графіком за нормальною гнучкою системою 8:00(10:00) - 17:00(19:00), отже, дорогою домою ЦА здійснює покупки, що є хорошим рушійним фактором для впровадження акцій та знижок на вечірні години, а також маркетингових кампаній щодо спеціальних пропозицій та тарифів за персональною карткою магазину із рекомендаційною системою відповідно до вподобань клієнта. </ul>
    <li><b>13:00</b></li>
    <ul style='font-family: Garamond ; font-size: 18px;'>Якщо висловити тезу про те, що обідня перерва здебільшого має період 12:00 - 13:00, то формулювання припущення/гіпотези наступне, що за lunch time, як правило, людина, сама володар своєму часу, перші 15-30 хвилин присвячує прийому їжі, а залишок - покупкам у магазині, щоб або перекусити протягом робочого дня, або скоротити дорогу додому, купивши всі необхідні речі під час обіду.</ul>
    <li><b>15:00</b></li>
    <ul style='font-family: Garamond ; font-size: 18px;'>Деяка корпоративна культура передбачає обідню перерву пізніше - 14:00 - 15:00, проте слід зауважити, що більшість компаній переходять до flexible schedule, тобто повне розпорядження робочим часом самостійно, окрім фіксованих мітів, постійного підтримку зв'язку та вчасно закритих дедлайнів. Припускаємо, що категорія покупців розраховують, що після обідніх годин-пік (13:00) магазини є менш наповненими. Такою стратегією мислення вони (покупці 15:00) створюють новий тренд, що відстає від попередньої позиції лише на 1 покупця.</ul>
    <li><b>10:00</b></li>
    <ul style='font-family: Garamond ; font-size: 18px;'>Ранкові покупці зазвичай включають пенсіонерів, домогосподарів, фрілансерів або людей, які прагнуть уникати натовпу. Оновлення та рефреш асортименту магазину дають клієнтам мотивацію приходити раніше.</ul>
    <li><b>18:00</b></li>
    <ul style='font-family: Garamond ; font-size: 18px;'>Агрументація щодо затребуваності товарів та високої активності за вище вказаної години тотожна до поясненню по 19:00.</ul>
</ul>
<hr></hr>
<p style='font-family: Garamond ; font-size: 18px;'><b>Крім того</b>, для більш детального аналізу слід продивитися окремо активність за святкових днів, вихідних та робочих, оскільки така аналітика дасть зрозуміти, за рахунок чого насправді формується загальна тенденція, тому data engineer'ові необхідно врахувати проєутвання колонки daytype для дослідження з перпективною розробки ефективних маркетингових стратегій не тільки на свята, а й певні тематичні будні, і ще краще - години, що є годинами-пік у нашому кейсі.</p>
<p style='font-family: Garamond ; font-size: 18px;'>Для сворення сильних стратегій продажів, беремо до уваги й окремий ресерч зайнятості населення міст, аби зрозуміти, на яких годинах і послугах роботи акцент. Якщо це постіндустрільне урбаністичне місто, то слід оптимізувати шляхи й види доставок + заохочувати окремими пропозиціями з мотивацією очного відвідування магазину. Робимо аналіз зайнятості суспільства: чіткий робочий графік, фріланс, змішаний формат, з метою розуміння, які бонуси та бенефіти спрацюють найкраще.</p>

5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

In [41]:
supermarket_dataset.groupby('City')['Total'].mean()

City
Mandalay     319.872506
Naypyitaw    337.099715
Yangon       312.354031
Name: Total, dtype: float64

6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

In [43]:
supermarket_dataset.groupby('City')['Total'].agg(['mean','min','max'])

mean      min      max
City                                   
Mandalay   319.872506  18.6375  1022.49
Naypyitaw  337.099715  10.6785  1042.65
Yangon     312.354031  12.6945  1039.29

<p style='font-family: Garamond ; font-size: 18px;'><b>Середні витрати покупців:</b></p>
<ul style='font-family: Garamond ; font-size: 18px;'>
    <li>Місто <b>Naypyitaw</b> має найвищий середній чек (337.10), що на 5.39% вище, ніж середній чек у <b>Mandalay</b> (319.87), і на 7.92% вище, ніж у <b>Yangon</b> (312.35).</li>
    <ul style='font-family: Garamond ; font-size: 18px;'>Це свідчить про вищу купівельну спроможність населення (ПКС - паритет купівельної спроможності) або наявність унікальних товарів із доданою вартістю. Уплив мають такой акційні пропозиції, що, можливо, найкраще та найефективніше спрацьовують на населення цього міста (наприклад, 1+1=3, де найдешевша одиниця товару йтиме у подарунок).</ul>
    <li><b>Yangon</b> демонструє найнижчий середній чек, що може бути індикатором більш низької платоспроможності або обмеженого доступу до преміум-продуктів.</li>
</ul>

<p style='font-family: Garamond ; font-size: 18px;'><b>Мінімальні витрати:</b></p>
<ul style='font-family: Garamond ; font-size: 18px;'>
    <li>Найменша вартість покупки (10.68) у <b>Naypyitaw</b> може вказувати на часті покупки недорогих товарів або доступність товарів першої необхідності.</li>
    <li>Мінімальні витрати у <b>Mandalay</b> (18.64) та <b>Yangon</b> (12.69) є значно вищими, що може свідчити про меншу кількість "дрібних" транзакцій у цих містах, а також більш сильні маркетингові та таргетовані кампанії товару люкс-сегменту.</li>
</ul>    

<p style='font-family: Garamond ; font-size: 18px;'><b>Максимальні витрати:</b></p>
<ul style='font-family: Garamond ; font-size: 18px;'>
    <li>Максимальні значення покупок <b>(≈1040)</b> є подібними для всіх міст. Це вказує на наявність окремої категорії покупців із високою платоспроможністю у всіх регіонах, що створює потенціал для преміум-пропозицій у кожному місті. Якщо розглядати окремо місто <b>Mandalay</b> як мінімальний максимум (<code>min(max)</code>) серед вибірки <b>(≈1020)</b> (проте мінімальний чек цього міста найвищий), то подальший більш глибокий аналіз буде зосереджено на таких метриках: аналіз асортименту, маркетингові дослідження, огляд впливу локацій, розробка стратегії для залучення клієнтів з високим доходом (причини: менша купівельна спроможність населення, менша кількість елітних товарів або дорогих послуг, культура покупок, конкуренція, клієнтський сегмент).</li>
</ul>

<hr></hr>
<p style='font-family: Garamond ; font-size: 18px;'><b>Рекомендації з точки зору оптимізації бізнес-процесів:</b></p>
<ol style='font-family: Garamond ; font-size: 18px;'>
    <li><b>Сегментація товарів і регіонів:</b></li>
    <ul style='font-family: Garamond ; font-size: 18px;'>
        <li>У Naypyitaw слід оптимізувати асортимент преміум-продуктів та товарів із високою маржею, оскільки високий середній чек демонструє попит на продукцію вищого сегмента.</li>
        <li>У Yangon, враховуючи нижчий середній чек, слід сфокусуватися на товарах середньої та нижньої цінових категорій. Це також вказує на потенціал для розвитку акційного та бюджетного сегмента.</li>
        <li>У Mandalay пропонується збалансована стратегія, орієнтована на підтримку середнього чека через знижки на масові товари та одночасне стимулювання преміум-покупок.</li>
    </ul>
    <li><b>Моделювання попиту:</b></li>
    <ul style='font-family: Garamond ; font-size: 18px;'>
        <li>Використовуючи дані середніх витрат, розробити економетричну модель прогнозування попиту з урахуванням специфіки кожного міста. Ця модель може включати фактори цінової еластичності попиту, сезонності та регіональні особливості.</li>
    </ul>
    <li><b>Рентабельність інвестицій:</b></li>
    <ul style='font-family: Garamond ; font-size: 18px;'>
        <li>Для Naypyitaw слід інвестувати в маркетинг та логістику преміум-сегмента з метою високого ROI (рентабельність інвестицій) через вищу купівельну спроможність.</li>
        <li>У Yangon оптимізація ціноутворення та логістики допоможе мінімізувати витрати на обслуговування клієнтів і збільшити рентабельність.</li>
    </ul>
    <li><b>Диверсифікація асортименту:</b></li>
    <ul style='font-family: Garamond ; font-size: 18px;'>
        <li>У Naypyitaw максимізувати дохід через преміум-продукцію, що приносить найбільший прибуток.</li>
        <li>У Mandalay і Yangon запропонувати більший асортимент доступних продуктів для збільшення кількості транзакцій і середнього чека.</li>
    </ul>
    <li><b>Оптимізація логістики:</b></li>
    <ul style='font-family: Garamond ; font-size: 18px;'>
        <li>Рівномірність максимальних витрат у всіх містах свідчить про потенціал для розвитку стратегій доставки дорогих товарів, які мають попит в усіх регіонах.</li>
    </ul>
    <li><b>Маркетингові стратегії та піар-експерименти</b> щодо підвищення попиту на ті товари, які є менш затребуваними у певних містах, шляхом упровадження акцій, знижок, тематичних годин тощо на відповідні позиції (та анаглогічні до них за рекомендаційними системами) щодо загальних потреб населення кожного регіону (additional analysis).</li>
</ol>    

7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.

In [47]:
supermarket_dataset.to_csv('supermarket_sales_enriched.csv', index=False)
supermarket_sales_enriched = pd.read_csv('supermarket_sales_enriched.csv')
supermarket_sales_enriched.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  \
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1   
1  10:29         Cash   76.40                 4.761905        3.8200     9.6   
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4   
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4   
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3   

  Rating Category  Hour  
0            High    13  
1            High    10  
2          Medium    13  
3            High    20  
4             Low    10